### ISSUE TITLE
ReMarkable beta 3.6: UnexpectedBlockError

### ISSUE THREAD

**Issue by Azeirah** [ricklupton/rmscene#14]
I'm getting this error in documents accessed with the ReMarkable beta v3.6. I haven't got a shareable file yet as I am not sure about how to reproduce it just yet, I will try to create a minimal example to reproduce it however.

This is the error I'm getting:

```py
Unknown block type 13. Skipping 31 bytes.
Traceback (most recent call last):
  File "/usr/local/lib/python3.10/runpy.py", line 196, in _run_module_as_main
    return _run_code(code, main_globals, None,
  File "/usr/local/lib/python3.10/runpy.py", line 86, in _run_code
    exec(code, run_globals)
  File "/app/remarks/__main__.py", line 151, in <module>
    main()
  File "/app/remarks/__main__.py", line 147, in main
    run_remarks(input_dir, output_dir, **args_dict)
  File "/app/remarks/remarks.py", line 96, in run_remarks
    process_document(metadata_path, out_path, doc_type, **kwargs)
  File "/app/remarks/remarks.py", line 188, in process_document
    pages_map.append(determine_document_dimensions(path))
  File "/app/remarks/conversion/parsing.py", line 183, in determine_document_dimensions
    build_tree(tree, blocks)
  File "/app/.venv/lib/python3.10/site-packages/rmscene/scene_stream.py", line 750, in build_tree
    for b in blocks:
  File "/app/.venv/lib/python3.10/site-packages/rmscene/scene_stream.py", line 725, in read_blocks
    yield from _read_blocks(stream)
  File "/app/.venv/lib/python3.10/site-packages/rmscene/scene_stream.py", line 707, in _read_blocks
    yield block_type.from_stream(stream)
  File "/app/.venv/lib/python3.10/site-packages/rmscene/scene_stream.py", line 376, in from_stream
    value = subclass.value_from_stream(stream)
  File "/app/.venv/lib/python3.10/site-packages/rmscene/scene_stream.py", line 468, in value_from_stream
    value = glyph_range_from_stream(reader)
  File "/app/.venv/lib/python3.10/site-packages/rmscene/scene_stream.py", line 420, in glyph_range_from_stream
    start = stream.read_int(2)
  File "/app/.venv/lib/python3.10/site-packages/rmscene/tagged_block_reader.py", line 88, in read_int
    self.data.read_tag(index, TagType.Byte4)
  File "/app/.venv/lib/python3.10/site-packages/rmscene/tagged_block_common.py", line 103, in read_tag
    raise UnexpectedBlockError(
    rmscene.tagged_block_common.UnexpectedBlockError: Expected index 2, got 4, at position 5992
```

From what I can tell, it looks like the glyph_range data has changed:

```py
def glyph_range_from_stream(stream: TaggedBlockReader) -> si.GlyphRange:
    start = stream.read_int(2)
    length = stream.read_int(3)
    color_id = stream.read_int(4)  # ddvk has this as a byte?
    color = si.PenColor(color_id)
    text = stream.read_string(5)
```

I can fix each error as it comes along, making the final code something like this, bruteforcing my way to the new struct, but I get stuck at PenColor where the Error tells me it's a Byte2, but I get a value in the 20000s as a pencolor when reading the TaggedBlock as a short.

```py
def glyph_range_from_stream(stream: TaggedBlockReader) -> si.GlyphRange:
    start = stream.read_int(4)
    length = stream.read_length(5) # has a tag 0xC == 12 bytes? Not sure what this is supposed to represent
    # i just bruteforced a function called read_length to skip the error but the returned data of course makes no sense.
    color_id = stream.read_short(0) 
    color = si.PenColor(color_id)
    text = stream.read_string(5)
```

---

Instead, I tried finding the struct definition using Ghidra. I'm stupendously inexperienced using tools like that so unfortunately I couldn't find much more than just references to the string `GlyphRange` :(

Do you have time to look at this? And if not, can you explain to me how to find the right values myself? It's been a while since I last did system programming in C/C++ but I'm not entirely unfamiliar with it either.

**Comment by ricklupton** [comment]
Hi @Azeirah, I don't know how to figure out the struct definition from the RM code, mostly when I wrote this I was following the structure that ddvk had already set out in his Go version as a starting point, and then just figuring things out from the .rm file contents...

I would next look at the binary .rm file data around the location mentioned in the error message. If you go to byte 5592 and then work backwards you should be able to see the start of the block.

In the example file `tests/data/Wikipedia_highlighted_p1.rm` there is a GlyphRange block that looks like this:

```
8a000000      # block length
00010103      # block header, last byte is 0x03 -> SceneGlyphItemBlock
1f 000b       # parent id, read in SceneItemBlock.from_stream
2f 0111       # item id
3f 0110       # left id
4f 0000       # right id
54 00000000   # deleted_length
6c 74000000   # start of subblock
01            # byte for item type.
              # following part is then read by subclass.value_from_stream, i.e. glyph_range_from_stream
24 00060000   # start
34 17000000   # length
44 03000000   # color id
5c 19000000   # start of subblock for string
17 01         # string header
6469 7370 6c61 7920 7465 6368 6e6f 6c6f 6779 2e5b 3133 5d   # string data
6c 41000000    # start of subblock
02             # number of rectangles
b2faa815 7e4889c0 0023086d 75f09a40 80365b14 afd87440 00807545 f4264c40    # rect 1 coords
e4b46344 2d587ec0 805216f7 be099b40 00c8cecd 2c1c4d40 00603e61 0cb64040    # rect 2 coords 
```

In your case, I would expect the position 5992 to be (roughly) at the point labelled "start" starting with `24`.

If you go back to find the start of the block and extract it similar to the above, and post here, we may be able to figure out what's changed in the format?

(if you don't want to share the text that has been highlighted, replace the "string data" part with "[XX bytes hidden]")

**Comment by Azeirah** [comment]
Ah thank you, I'll go take a look coming Wednesday. I also got the book about Ghidra ddvk recommended in the ReMarkable discord server so I have that as a backup as well.

The person who had this error said there is no text on the page, so I'm curious to see what I'll find.

**Comment by Azeirah** [comment]
Ok, I took a look at the file. There's clearly some text in the document, but it's deleted. It's not sensitive, the person was probably just testing or accidentally typing something.

The raw data at the relevant addresses is, I believe it starts at 00 6300 00. I determined the start by working backwards from the block header with byte 0x03.

```
000017a0:  0101 2b91 0100 6300 0000 0001 0103 1f00 0b2f 019f 063f 019e
000017b8:  064f 0000 5400 0000 006c 4b00 0000 0144 0500 0000 5c10 0000
000017d0:  000e 0171 7764 6672 6767 6763 2020 2020 676c 2100 0000 0100
```

Here's my attempt at hand-parsing the block

```
## These are all the same
00630000 			# block length = 
00000103 			# block header
1f  000b 			# parent id
2f  019f			# item id
3f  019e06			# left id    			--- a byte longer than usual?
4f  0000			# right id
54  0000006c		# deleted length
4b  00000001		# start of subblock
# From here on it changes
# The raw data is:
#										string data
# 							  vv vvvv vvvv vvvv vvvv vvvv vvvv vvvv vvvv
44 0500 0000 5c10 0000 000e 0171 7764 6672 6767 6763 2020 2020 676c 2100 0000 0100
#																		 ^^^^ ^^^^
# 																		 next block?

My guess at parsing this is the following:
44 0500 0000 		# color id = 5
5c 1000 0000 		# start of subblock for string
0e 01	     		# string header?
```

**Comment by Azeirah** [comment]
Woah I was thinking in completely the wrong way, GlyphRange is not text written with the text tool or converted with the OCR tool, it's highlighted text!

I believe I found the correct encoding

```py
def glyph_range_from_stream(stream: TaggedBlockReader) -> si.GlyphRange:
    color_id = stream.read_int(4)  # ddvk has this as a byte?
    color = si.PenColor(color_id)
    text = stream.read_string(5)
    length = len(text)
    start = None
```

Start is gone, length is still in there, it's right before the `0x01` string header.

For example, `GlyphRange` with `"hello world"` in a different and new document

```
000001b0:  0000 5400 0000 006c 4800 0000 0144 0000 0000 5c0d 0000 000b  :..T....lH....D....\.....
000001c8:  0168 656c 6c6f 2077 6f72 6c64 6c21 0000 0001 0000 0000 00f0  :.hello worldl!..........
```

The string starts exactly at the second line `0168` where `01` is the header, `68` is the `h`. `The 0b` is the string length = 11 which matches the actual string length of `"hello world"`. This length is already read in your `read_string` implementation.

I think all they did was get rid of their `start` and `length` fields to save some space. Start is probably just not necessary whatsoever and the length is already present in the string data.

**Comment by Azeirah** [comment]
The only thing I don't know yet is how do I do this conditionally based on a version? I saw min_version and current_version but I'm not sure how that relates to the `GlyphRange` yet. The structure of the document is still very confusing.

There are Blocks and subblocks, I see that for instance the `SceneLineItemBlock` has a version, but `SceneGlyphItemBlock` does not?

**Comment by ricklupton** [comment]
> Ok, I took a look at the file. There's clearly some text in the document, but it's deleted. It's not sensitive, the person was probably just testing or accidentally typing something as the text is `"qwdfrgggc gl!"`. Length = 16, deleted_length = 16 in the parsed block so it makes sense.
> 
> Here's the relevant raw data. I believe it starts at [00 6300 00](tel:00 6300 00). I determined the start by working backwards from the block header with byte 0x03:
> 
> ```
> Start and end is marked.
>                           V
> 000017a0:  0101 2b91 0100 6300 0000 0001 0103 1f00 0b2f 019f 063f 019e
> 000017b8:  064f 0000 5400 0000 006c 4b00 0000 0144 0500 0000 5c10 0000
> 000017d0:  000e 0171 7764 6672 6767 6763 2020 2020 676c 2100 0000 0100
>                                                           ^
> ```
> 
> Here's my attempt at hand-parsing the block
> 
> ```
> ## These are all the same
> 00630000             # block length
> 00000103             # block header
> 1f  000b             # parent id
> 2f  019f06           # item id   --- a byte longer than usual?
> 3f  019e06           # left id   --- a byte longer than usual?
> 4f  0000             # right id
> 54  0000006c         # deleted length=16
> 4b  00000001         # start of subblock

I think you missed a 00 here? The deleted length is 0 (which is usually the case for an item which does have content, like this one). Then 6c introduces a subblock with index 6, whose length is 0x0000004b. 

> # From here on it changes
> # The raw data is:
> #					string data
> # 			      vv vvvv vvvv vvvv vvvv vvvv vvvv vvvv vvvv
> 44 0500 0000 5c10 0000 000e 0171 7764 6672 6767 6763 2020 2020 676c 2100 0000 0100
> ```
> 
> It's clear there's a lot less data up until the string start included there. I can see the `44` and the `5c` in there like in your pasted contents. Those are the `color_id` and `start of subblock` in your example. My attempt to parse it looks like this:
> 
> ```
> 44 0500 0000     # color id = 5
> 5c 1000 0000     # start of subblock
> 0e 01            # string header?
> ```
> 
> `TagType.Length4 = 0xC`, the `start of subblock` tag is `5c`, Perhaps it should read
> 
> `5c 1000 0000 0e01`?
> 
> And it is implicitly assumed that when the block ends the string immediately starts? Or perhaps the 0e and 01 have special meaning and need to be parsed invididually despite being tagged.

As you say below, the 0x0e is the string length, and 0x01 is some kind of flag which always seems to be 01. 

> 
> Can you elaborate on what the `Length4` Tag is meant for? ID, Byte8, Byte4 and Byte2 all make sense to me but not this one as it doesn't directly map to an elementary datatype like i64 or i8 (rust notation)

“Length4” is the Tag type used for introducing subblocks — Length refers to the length of the subsequent subblock. 

> 
> ## Block
> I noticed the item_id and left_id are both one byte longer? That makes 24 bits which is odd? I suppose this is still ok, since `read_crdt_id` calls `part2 = self.read_varuint()` so the length of the second part is variable.
> 
> ```python
>     def read_crdt_id(self) -> CrdtId:
>         # Based on ddvk's reader.go
>         # TODO: should be var unit?
>         part1 = self.read_uint8()
>         part2 = self.read_varuint()
>         # result = (part1 << 48) | part2
>         return CrdtId(part1, part2)
> ```
> 

Yes, that’s right

> ## GlyphRange
> * `start` is missing.
> * `length` is missing. The byte after the last character in the string _is_ a 0x00 so perhaps they changed it to be null-terminated? Not improbable
> * `color_id` is still present, in this text it would be `5`
> * `start of subblock` is also still present
> * `string header` has changed? 0e 01. I'm not entirely sure what a string header "is", strings in C don't have a start-byte, right? Regardless, it still ends with 01. I'm not sure what the `0e` might mean.

**Comment by ricklupton** [comment]
> Woah I was thinking in completely the wrong way, GlyphRange is not text written with the text tool or converted with the OCR tool, it's highlighted text!

Ah yes, that’s right!

> 
> I believe I found the correct encoding
> 
> ```python
> def glyph_range_from_stream(stream: TaggedBlockReader) -> si.GlyphRange:
>     color_id = stream.read_int(4)  # ddvk has this as a byte?
>     color = si.PenColor(color_id)
>     text = stream.read_string(5)
>     length = len(text)
>     start = -1
> ```
> 
> Start is gone, length is still in there, it's right before the `0x01` string header.
> 
> For example, `GlyphRange` with `"hello world"` in a different and new document
> 
> ```
> 000001b0:  0000 5400 0000 006c 4800 0000 0144 0000 0000 5c0d 0000 000b  :..T....lH....D....\.....
> 000001c8:  0168 656c 6c6f 2077 6f72 6c64 6c21 0000 0001 0000 0000 00f0  :.hello worldl!..........
> ```
> 
> The string starts exactly at the second line `0168` where `01` is the header, `68` is the `h`. `The 0b` is the string length = 11 which matches the actual string length of `"hello world"`. This length is already read in your `read_string` implementation.
> 
> I think all they did was get rid of their `start` and `length` fields to save some space. Start is probably just not necessary whatsoever and the length is already present in the string data.

That makes sense, and a clear example. Although, I think when testing I did find some cases where the string length was not the same as the highlighted length, possibly when the highlighted characters had accents and took multiple bytes to store (I didn’t figure out exactly why, or how the start position was measured). Maybe they decided it was ok to just ignore this and keep the actual highlighted text and location on the page.

**Comment by Azeirah** [comment]
> possibly when the highlighted characters had accents and took multiple bytes to store

Good point, I hadn't considered non-ascii (there are barely any non-ascii characters in Dutch nor in English) characters but the strings are of course utf-8 given that they support all kinds of languages.

**Comment by ricklupton** [comment]
> The only thing I don't know yet is how do I do this conditionally based on a version? I saw min_version and current_version but I'm not sure how that relates to the `GlyphRange` yet. The structure of the document is still very confusing.
> 
> There are SceneBlocks, and subblocks, I see that for instance the `SceneLineItemBlock` has a version, but `SceneGlyphItemBlock` does not?
> 
> Printing the (updated with my parsing logic) block gives me this
> 
> ```
> GlyphRange(start=-1, length=11, text='hello world', color=<PenColor.BLACK: 0>, rectangles=[Rectangle(x=-479.0, y=223.0, w=345.609375, h=90.0)])
> ```

I think it can still be read without needing an explicit version number. Each piece of data has an “index” number. The two pieces that have been removed, the start and the length, had indexes 1 and 2. The next piece, which is still there, still has index 3. So, the parsing code needs updating to try to read indices 1 and 2, but if they are not present, to fall back on your version. 

Minor thing but probably using None rather than -1 would be clearer for a missing start position?

**Comment by Azeirah** [comment]
> I think it can still be read without needing an explicit version number.

I personally prefer being explicit about these kinds of things, but it's your call since it's your repo ;p

> Minor thing but probably using None rather than -1 would be clearer for a missing start position?

Agreed, it's a bad habit I took from our legacy PHP codebase at work

**Comment by Azeirah** [comment]
I changed the typing to Optional[int] for the start field.

**Comment by Azeirah** [comment]
From what I found in another document, it looks like the version number does not correspond one-to-one to the new format. So your proposed approach of try-catching works better.

I think the format differs depending on whether the highlight is deleted or not. Given that a deleted highlight can't really have a start. It might also be that the text under the highlight itself got deleted later, although that sounds a little more complicated.

**Comment by ricklupton** [comment]
I think this is closed by #15
